src/test/java/com/google/devtools/build/lib/actions/ActionCacheCheckerTest.java

```diff
@@ -391,6 +432,135 @@ public void testDeletedConstantMetadataOutputCausesReexecution() throws Exceptio
+            artifactRoot, PathFragment.create("bin/dummy"), NULL_ARTIFACT_OWNER);
+    String content = "content";
+    Action action = new InjectOutputMetadataAction(output, createRemoteMetadata(content));
+
+    // act
+
+    // Not cached.
+    runAction(action);
+
+    // assert
+    assertThat(output.getPath().exists()).isFalse();
+    ActionCache.Entry entry = cache.get(output.getExecPathString());
+    assertThat(entry).isNotNull();
+    assertThat(entry.getOutputFile(output)).isEqualTo(createRemoteMetadata(content));
+    assertStatistics(0, new MissDetailsBuilder().set(MissReason.NOT_CACHED, 1).build());
+  }
+
+  @Test
+  public void saveOutputMetadata_localFileMetadataNotSaved() throws Exception {
+    // arrange
+    enableSaveOutputMetadata();
+
+    Artifact.DerivedArtifact output =
+        new Artifact.DerivedArtifact(
+            artifactRoot, PathFragment.create("bin/dummy"), NULL_ARTIFACT_OWNER);
+    output.getPath().getParentDirectory().createDirectoryAndParents();
+    FileSystemUtils.writeContentAsLatin1(output.getPath(), "content");
+    Action action = new InjectOutputMetadataAction(output, FileArtifactValue.createForTesting(output));
+    output.getPath().delete();
+
+    // act
+
+    // Not cached.
+    runAction(action);
+
+    // assert
+    assertThat(output.getPath().exists()).isFalse();
+    ActionCache.Entry entry = cache.get(output.getExecPathString());
+    assertThat(entry).isNotNull();
+    assertThat(entry.getOutputFile(output)).isNull();
+    assertStatistics(0, new MissDetailsBuilder().set(MissReason.NOT_CACHED, 1).build());
+  }
+
+  @Test
+  public void saveOutputMetadata_notSavedIfNotEnabled() throws Exception {
+    // arrange
+    Artifact.DerivedArtifact output =
+        new Artifact.DerivedArtifact(
+            artifactRoot, PathFragment.create("bin/dummy"), NULL_ARTIFACT_OWNER);
+    String content = "content";
+    Action action = new InjectOutputMetadataAction(output, createRemoteMetadata(content));
+
+    // act
+
+    // Not cached.
+    runAction(action);
+
+    // assert
+    assertThat(output.getPath().exists()).isFalse();
+    ActionCache.Entry entry = cache.get(output.getExecPathString());
+    assertThat(entry).isNotNull();
+    assertThat(entry.getOutputFile(output)).isNull();
+    assertStatistics(0, new MissDetailsBuilder().set(MissReason.NOT_CACHED, 1).build());
+  }
+
+  @Test
+  public void saveOutputMetadata_remoteFileMetadataLoaded() throws Exception {
+    // arrange
+    enableSaveOutputMetadata();
+
+    Artifact.DerivedArtifact output =
+        new Artifact.DerivedArtifact(
+            artifactRoot, PathFragment.create("bin/dummy"), NULL_ARTIFACT_OWNER);
+    String content = "content";
+    Action action = new InjectOutputMetadataAction(output, createRemoteMetadata(content));
+
+    // act
+
+    // Not cached.
+    runAction(action);
+    // Cache hit
+    runAction(action);
+
+    // assert
+    assertThat(output.getPath().exists()).isFalse();
```

Comment:
Should we assert the output metadata in the entry? Also, how about checking whether the outputs were injected to the metadata handler?